### PR TITLE
Use pre-commits

### PR DESCRIPTION
**Pull request by thomasmfish**
I looked into using these for the docs but it gets incredibly messy with environments or slow trying to get the python help text, but these options looked useful.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,36 @@
+repos:
+  - repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: v4.6.0
+    hooks:
+      - id: check-yaml
+      - id: end-of-file-fixer
+      - id: trailing-whitespace
+
+  - repo: https://github.com/pycqa/flake8
```
Comment (PeterC-DLS) [inline .pre-commit-config.yaml]: Can't you replace flake8 and its plugins by ruff alone with its rules

Comment (thomasmfish) [reply .pre-commit-config.yaml]: Ah, yep

Comment (PeterC-DLS) [reply .pre-commit-config.yaml]: I thought you would add some ruff settings to select the removed plugins' rules

Comment (thomasmfish) [reply .pre-commit-config.yaml]: I'm fine with the defaults, honestly

**Review by PeterC-DLS — APPROVED** [review]

**Pull request merged**